pkg/download/consistent_hashing.go

```diff
@@ -0,0 +1,263 @@
+		Int("concurrency", m.maxConcurrency()).
+		Ints64("chunks_per_slice", chunksPerSlice).
+		Msg("Downloading")
+
+	for slice, numChunks := range chunksPerSlice {
+		if numChunks == 0 {
+			// this happens if we've already downloaded the whole first slice
+			continue
+		}
+		start := m.SliceSize * int64(slice)
+		sliceSize := m.SliceSize
+		if slice == 0 {
+			start = firstChunkResp.ContentLength
+			sliceSize = sliceSize - firstChunkResp.ContentLength
+		}
+		if slice == int(totalSlices)-1 {
+			sliceSize = (fileSize-1)%m.SliceSize + 1
+		}
+		if sliceSize/numChunks < m.minChunkSize() {
+			// reset numChunks to respect minChunkSize
+			numChunks = sliceSize / m.minChunkSize()
+			// although we must always have at least one chunk
+			if numChunks == 0 {
+				numChunks = 1
+			}
+		}
+		chunkSizes := EqualSplit(sliceSize, numChunks)
+		for _, chunkSize := range chunkSizes {
+			end := start + chunkSize - 1
+
+			logger.Debug().Int64("start", start).Int64("end", end).Msg("starting request")
+			resp, err := m.doRequest(ctx, start, end, urlString)
+			if err != nil {
+				return nil, -1, err
+			}
+
+			dataSlice := data[start : end+1]
+			errGroup.Go(func() error {
+				return m.downloadChunk(resp, dataSlice)
+			})
+
+			start = start + chunkSize
+		}
+	}
+
+	if err := errGroup.Wait(); err != nil {
+		return nil, -1, err // return the first error we encounter
+	}
+
+	buffer := bytes.NewBuffer(data)
+	return buffer, fileSize, nil
+}
+
+func (m *ConsistentHashingMode) doRequest(ctx context.Context, start, end int64, urlString string) (*http.Response, error) {
+	logger := logging.GetLogger()
+	if m.Semaphore != nil {
+		err := m.Semaphore.Acquire(ctx, 1)
+		if err != nil {
+			return nil, err
+		}
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", urlString, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to download %s: %w", req.URL.String(), err)
+	}
+	err = m.consistentHashIfNeeded(req, start, end)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", start, end))
+
+	logger.Debug().Str("url", urlString).Str("munged_url", req.URL.String()).Str("host", req.Host).Int64("start", start).Int64("end", end).Msg("request")
+
+	resp, err := m.Client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error executing request for %s: %w", req.URL.String(), err)
+	}
+	if resp.StatusCode == 0 || resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("unexpected http status downloading %s: %s", req.URL.String(), resp.Status)
+	}
+
+	return resp, nil
+}
+
+func (m *ConsistentHashingMode) consistentHashIfNeeded(req *http.Request, start int64, end int64) error {
+	logger := logging.GetLogger()
+	for _, host := range m.DomainsToCache {
+		if host == req.URL.Host {
+			if start/m.SliceSize != end/m.SliceSize {
+				return fmt.Errorf("Can't make a range request across a slice boundary: %d-%d straddles a slice boundary (slice size is %d)", start, end, m.SliceSize)
+			}
```

Comment:
Is this mostly suspenders and belt? We should be able to avoid this scenario in all cases with something like:
minChunkSize > SliceSize -> minChunkSize == SliceSize -> chunkSize = SliceSize

SliceSize > minChunkSize > SliceSize/2 -> Round up or RoundDown? Preference roundDown and know we'll just use 2 chunks per slice always.

SliceSize % minChunkSize != 0, and does not fall into the above buckets -- we already handle this by having a smaller final chunk


To be clear, this code is fine as is. Thinking about how to make it more robust long term.

Reply:
yeah this is a "can't happen" guard